### PR TITLE
REFACTOR-#6812: Remove 'PyarrowOnRay' execution in favour of pyarrow-backed pandas dataframes

### LINKED ISSUE #6812: Consider to remove experimental `PyarrowOnRay` execution
Some details:
* Pandas supports `pyarrow` as a backend for its objects, starting with version 2.0 (optional). Modin can reuse it for different engines (Dask, Ray, Unidist).
* Moreover, it is [planned](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#pyarrow-will-become-a-required-dependency-with-pandas-3-0) to make `pyarrow` dependency required starting with version 3.0.
* Current testing of this execution is extremely sparse. At CI we ran only `TestCsv` tests (we're [skipping](https://github.com/modin-project/modin/blob/1a1e090032a187409515ce5697156a0716c4a599/modin/pandas/test/test_io.py#L257) even those tests right now).

@devin-petersohn your feedback here is especially valuable, as I recall it was implemented before many of us started developing Modin.

@modin-project/modin-core your feedback is also welcome

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6812 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by YarShev** [comment]
There is references to pyarrow.gandiva in the code. Should we remove it too?

File: docs/development/architecture.rst [hunk 1]
```diff
@@ -56,7 +56,7 @@ For the simplicity the other execution systems - Dask and MPI are omitted and on
   on a selected storage format and mapping or compiling the Dataframe Algebra DAG to and actual
```
Comment (YarShev) [inline docs/development/architecture.rst:56]: The line `.. _Arrow tables: https://arrow.apache.org/docs/python/generated/pyarrow.Table.html` should be remove in this file.

**Comment by anmyachev** [comment]
> There are references to pyarrow.gandiva in the code in doc_checker and docs/conf.py. Should we remove it too?

I removed it, except for `troubleshooting` section, since users can still use `pyarrow.gandiva` with Modin on HDK.

File: docs/development/architecture.rst (hunk 1 above)
Comment (anmyachev) [reply docs/development/architecture.rst:56]: removed

**Pull request merged**